File: tests/matlab/test_numdim.m
function tests = test_numdim
% TEST_NUMDIM  Tests for numdim.m (effective number of sources via
% eigenvalue-entropy of the channel second-order matrix A*A'/100).
%
% Expectations are closed-form / release-invariant:
%   - single channel             -> lambda == 1
%   - orthogonal equal-energy    -> lambda == nchan (uniform eigenvalues)
%   - identical channels (rank-1) -> lambda == 1 (one dominant eigenvalue)
%   - full-rank                  -> real, 1 < lambda < nchan
tests = functiontests(localfunctions);
end

function testSingleChannel(testCase)
% One channel -> one (normalized) eigenvalue == 1 -> entropy 0 -> lambda 1.
a = rand(1, 50) + 0.5;            % nonzero energy
verifyEqual(testCase, numdim(a), 1, 'AbsTol', 1e-10);
end

function testTwoChannelOrthogonal(testCase)
% A*A' = 2*I -> equal eigenvalues -> lambda == nchan == 2.
A = [1 1; 1 -1];
verifyEqual(testCase, numdim(A), 2, 'AbsTol', 1e-10);
end

function testOrthogonalEqualEnergy(testCase)
% Hadamard rows are orthogonal with equal norm -> A*A' = 4*I -> lambda == 4.
A = hadamard(4);
verifyEqual(testCase, numdim(A), 4, 'AbsTol', 1e-9);
end

function testRankDeficientApproxOne(testCase)
% Identical channels -> rank-1 -> one dominant eigenvalue -> ~1 effective dim
% (eig yields tiny nonzero eigenvalues, so the result is finite, not NaN).
A = ones(3, 10);
verifyEqual(testCase, numdim(A), 1, 'AbsTol', 1e-6);
end

function testFullRankBounds(testCase)
% Full-rank random data: real scalar strictly between 1 and nchan.
rng(42);
A = randn(5, 200);
v = numdim(A);
verifyTrue(testCase, isreal(v));
verifyGreaterThan(testCase, v, 1);
verifyLessThan(testCase, v, 5);
end

function testNoArgShowsHelp(testCase)
% nargin<1 branch: prints help and returns without error.
out = evalc('numdim()');
verifyTrue(testCase, ~isempty(out));
end

File: tests/test_numdim.py
"""Parity tests for numdim - effective number of sources (EEGLAB numdim.m).

numdim estimates a lower bound on the number of discrete sources via the
eigenvalue entropy of the channel second-order matrix ``A @ A.T / 100``. The
measure is scale-invariant (eigenvalues are normalized), so expectations below
are closed-form / property-based.

Expected values match the MATLAB reference tests in tests/matlab/test_numdim.m
(confirmed on MATLAB R2025a). Regenerate ground truth with
``matlab -batch "addpath('.../functions/miscfunc'); numdim(<input>)"``.
"""

from __future__ import annotations

import numpy as np
import pytest

from eegprep.functions.miscfunc.numdim import numdim
from tests.eeglab_tests import assert_matlab_near, eeglab_test

pytestmark = pytest.mark.parity


@eeglab_test("unittesting_miscfunc/numdim/miscfunc_numdim_wrapperTest.m", "test_pass_general")
@eeglab_test("unittesting_miscfunc/numdim/miscfunc_numdim_wrapperTest.m", "test_pass_row_vector")
@eeglab_test("unittesting_miscfunc/numdim/miscfunc_numdim_wrapperTest.m", "test_pass_column_vector")
@eeglab_test("unittesting_miscfunc/numdim/miscfunc_numdim_wrapperTest.m", "test_pass_singular_matrix")
def test_reference_numdim(eeglab_backend):
    result = eeglab_backend("numdim", np.array([[2, 1], [-1, -2]], dtype=float))
    assert_matlab_near(result, [[1 / (0.1**0.1) / (0.9**0.9)]])
    vector = np.array([[1, 2, 0, -6]], dtype=float)
    assert_matlab_near(eeglab_backend("numdim", vector), [[1]])
    result = eeglab_backend("numdim", vector.T)
    # Both alternatives are explicitly allowed by the MATLAB source test.
    assert_matlab_near(result, [[np.nan]] if np.isnan(result).all() else [[1]])
    eeglab_backend("numdim", np.array([[2, 2], [-1, -1]], dtype=float))


def test_single_channel_is_one():
    # One channel -> one normalized eigenvalue -> entropy 0 -> lambda == 1.
    rng = np.random.default_rng(0)
    a = rng.random((1, 50)) + 0.5
    np.testing.assert_allclose(numdim(a), 1.0, atol=1e-10)


def test_two_channel_orthogonal_is_nchan():
    # A @ A.T = 2*I -> equal eigenvalues -> lambda == nchan == 2.
    a = np.array([[1.0, 1.0], [1.0, -1.0]])
    np.testing.assert_allclose(numdim(a), 2.0, atol=1e-10)


def test_hadamard_equal_energy_is_nchan():
    # Hadamard(4): orthogonal equal-norm rows -> A @ A.T = 4*I -> lambda == 4.
    h4 = np.array(
        [
            [1, 1, 1, 1],
            [1, -1, 1, -1],
            [1, 1, -1, -1],
            [1, -1, -1, 1],
        ],
        dtype=float,
    )
    np.testing.assert_allclose(numdim(h4), 4.0, atol=1e-9)


def test_rank_deficient_is_approx_one():
    # Identical channels -> rank-1 -> one dominant eigenvalue -> ~1 effective dim.
    a = np.ones((3, 10))
    np.testing.assert_allclose(numdim(a), 1.0, atol=1e-6)


def test_full_rank_matches_matlab():
    # Full-rank deterministic integer matrix: identical in numpy and MATLAB, so
    # pin to the exact MATLAB R2025a value (confirmed via matlab -batch). The
    # result is real and strictly between 1 and nchan (=3).
    a = np.array(
        [
            [2.0, 1.0, 0.0, 1.0],
            [1.0, 3.0, 1.0, 0.0],
            [0.0, 1.0, 2.0, 1.0],
        ]
    )
    v = numdim(a)
    assert np.isreal(v)
    np.testing.assert_allclose(v, 2.147746217856, rtol=1e-6)


def test_python_regression_matches_closed_form_entropy():
    data = np.array([[2.0, 1.0], [-1.0, -2.0]])
    expected = 1 / (0.1**0.1) / (0.9**0.9)

    np.testing.assert_allclose(numdim(data), expected, rtol=1e-12)


def test_python_regression_row_vector_is_one_dimensional():
    assert numdim(np.array([[1.0, 2.0, 0.0, -6.0]])) == pytest.approx(1.0)


def test_python_regression_preserves_column_vector_degeneracy():
    # The upstream test allows NaN or 1 because numdim.m evaluates 0*log(0)
    # literally. The analytic entropy limit is one effective source.
    result = numdim(np.array([[1.0], [2.0], [0.0], [-6.0]]))

    assert result == pytest.approx(1.0)


def test_python_regression_singular_matrix_scenario_is_executable():
    # The analytic entropy of a rank-one matrix is one.
    result = numdim(np.array([[2.0, 2.0], [-1.0, -1.0]]))

    assert result == pytest.approx(1.0)
